zigzag-circuit.md

```diff
@@ -0,0 +1,122 @@
+- Inclusion Proof: For each inclusion path in the public inputs, we provide a Merkle Tree path
+  - `InclusionHash_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}_{0..TREE_DEPTH-1}`
+  - `ReplicaInclusionHash_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}_{0..TREE_DEPTH-1}`
+  - `ParentInclusionHash_{l=0..LAYERS}_{c=0..LAYER_CHALLENGES[i]}_{0..EXPANSION_DEGREE+ BASE_DEGREE}_{0..TREE_DEPTH-1}`
+  - `InclusionLeaf_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}`
+  - `ReplicaInclusionLeaf_{i=0..LAYERS}_{0..LAYER_CHALLENGES[i]}`
+  - `ParentInclusionLeaf_{l=0..LAYERS}_{c=0..LAYER_CHALLENGES[i]}_{0..EXPANSION_DEGREE+ BASE_DEGREE}`
+
+**Circuit:**
+
+- **Check** `ReplicaId` is equal to its bit representation
+
+  ```
+  Assign replica_id_bits = Fr_to_bits(ReplicaId)
+  Check Packed(replica_id_bits) == ReplicaId
+  ```
+
+- For each `l = 0..LAYERS`:
+
+  - For each `c = 0..LAYERS_CHALLENGES[l]`
+
+    - Inclusion checks:
+
+      - Correct inclusions proofs: **Check**  that all the inclusion proofs are correct
+
+        ```
+        Check MerkleTreeVerify(InclusionHash_{l}_{c}_{0..TREE_DEPTH-1})
+        Check MerkleTreeVerify(ReplicaInclusionHash_{l}_{c}_{0..TREE_DEPTH-1})
+        
+        For p = 0..(EXPANSION_DEGREE + BASE_DEGREE):
+        	Check MerkleTreeVerify(ParentInclusionHash_{l}_{c}_{p}_{0..TREE_DEPTH-1})
+        ```
+
+      - Correct layer: **Check** that `CommR_{l}` is matching the Replica Inclusion proofs root hash and `CommR_{l-1}` is matching the Inclusion proof root hash. (If `l=0`, use `CommD` instead, if `l=LAYER-1` use `CommRLast` instead).
+
+        ```
+        Check CommR_{l-1} === InclusionHash_{l}_{c}_{0}
+        Check CommR_{l} === ReplicationInclusionHash_{l}_{c}_{0}
+        
+        For parent = 0..(EXPANSION_DEGREE + BASE_DEGREE):
+        	Check CommR_{l} === ParentInclusionHash_{l}_{c}_{parent}_{0}
+        ```
+
+    - Encoding checks: **Check** that a challenged replica node decodes to the correct data node.
+
+      - **Check** that each parent has a correct bit representation:
+
+        ```
+        For each parent = ParentInclusionLeaf_{l}_{c}_{p}`:
+          Assign ParentBits_{l}_{c}_{p} = Fr_to_bits(parent)
+          Check Packed(ParentBits_{l}_{c}_{p}) == parent
+        ```
+
+      - **Check** that the KDF was run correctly:
+
+        ```
+        Assign leaf_bits = ParentBits_{l}_{c}_{0} || .. || ParentBits_{l}_{c}_{EXPANSION_DEGREE+BASE_DEGREE}
+        
+        Assign key : Fr = KDF(replica_id_bits, leaf_bits)
+        Check KDF(replica_id_bits, leaf_bits)
+        ```
+
+      - **Check** correct encoding: the decoded leaf equals the input data leaf
+
+        ```
+        Check ReplicationInclusionValue_{l}_{c} == InclusionValue_{l}_{c} + k
+        ```
+
+- CommRStar check: **Check** that CommRStar is computed by concatenating the replica identifier and the commR at each layer (specified in the aux inputs)
+
+  ```
+  Check CommRStar == PedersenHash(ReplicaId || CommR_{0} || .. || CommR_{LAYERS-1} || CommRLast)
```

Comment:
aren't `CommR_{LAYERS-1}` and `CommRLast` the same thing?

Reply:
that should be LAYERS-2, thank you!